### ISSUE TITLE
[ENH]: Sphinx plot_directive: add an option for embedding pngs as base64 data-urls

### ISSUE THREAD

**Issue by martinRenou** [matplotlib/matplotlib#21918]
### Problem

When using the Sphinx `plot_directive` in a context like IPython + JupyterLab, pointing to image files does not work, as showed in the following screenshot:

![jupyterlab_sphinxify](https://user-images.githubusercontent.com/21197331/145603984-1113b111-934f-4b4c-acc5-0edcd6f7d2da.png)

### Proposed solution

It would be nice to add an option for embedding pngs as base64 data in the resulting HTML.

If you're open to such a feature, I can make a PR myself adding this option.

**Comment by timhoffm** [comment]
Where does the html come from?

**Comment by QuLogic** [comment]
I would not expect that code to actually be run. I guess this is a thing that `sphinxify_docstring` does, but then isn't this more of a bug for them?

**Comment by martinRenou** [comment]
> Where does the html come from?

The HTML comes from this jinja template in the plot directive:

https://github.com/matplotlib/matplotlib/blob/main/lib/matplotlib/sphinxext/plot_directive.py#L361

> I would not expect that code to actually be run

I understand this code is run here: https://github.com/matplotlib/matplotlib/blob/main/lib/matplotlib/sphinxext/plot_directive.py#L517

Then the plot directive gathers figure from the manager and save them into separate files: https://github.com/matplotlib/matplotlib/blob/main/lib/matplotlib/sphinxext/plot_directive.py#L636

**Comment by QuLogic** [comment]
The plot directive is run by Sphinx; I mean, what is running Sphinx?

**Comment by martinRenou** [comment]
> what is running Sphinx?

Sorry, I thought you were talking about the code in the docstring.

By setting `InteractiveShell.sphinxify_docstring` to True in IPython, the code inspection will use docrepr's sphinxify function which runs sphinx to generate the HTML output.

See:
- https://github.com/ipython/ipython/blob/master/IPython/core/interactiveshell.py#L341
- https://github.com/ipython/ipython/blob/master/IPython/core/interactiveshell.py#L1624
- https://github.com/ipython/ipython/blob/master/IPython/core/interactiveshell.py#L91-L98

(If you look at the docrepr's codebase, you'll see that it actually uses an old copy of the `plot_directive`, but this will [hopefully be fixed soon](https://github.com/spyder-ide/docrepr/pull/30))

It runs sphinx in a temporary folder that the JupyterLab server is not serving, and IPython [only outputs the HTML result](https://github.com/ipython/ipython/blob/master/IPython/core/interactiveshell.py#L96). The plot image being in this temp folder, it is nowhere to be found for the server.

Now back to the original issue: I would like to suggest an option for the `plot_directive` that would allow embedding images as base64 in the HTML output, so that it doesn't need to be served.

**Comment by timhoffm** [comment]
I see the use case for embedding. However, I don't think it should be a simple option on `plot_directive`. We don't want the images embedded in our regular HTML docs. This would need to be a run-time switch.

Then again, couldn't sphinxify postprocess the HTML and embed the images themselves? I have the feeling it's mostly their  concern, and the issue seems to apply to all images, not only plots.

**Comment by martinRenou** [comment]
> We don't want the images embedded in our regular HTML docs. This would need to be a run-time switch

Yeah definitely. I was under the impression that it would be possible to put such feature behind a flag and disable this by default.

> the issue seems to apply to all images, not only plots.

For sure.

In that sense it might be more sensible to change docrepr directly indeed. Though my only concern is that the `plot_directive` is the one responsible for generating the output HTML, so I am not sure it could be implemented without a hacky solution that goes through the output HTML and look for images to make them embedded.

**Comment by timhoffm** [comment]
I‘m closing this as it’s better be handled upstream.
